presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java

```diff
@@ -186,55 +174,29 @@ protected ColumnStatistics createColumnStatistics()
         return statistics;
     }
 
-    private static int[] getSortedDictionaryNullsLast(Block elementBlock)
+    private static int[] getSortedDictionary(DictionaryBuilder dictionary)
     {
-        int[] sortedPositions = new int[elementBlock.getPositionCount()];
+        int[] sortedPositions = new int[dictionary.getEntryCount()];
         for (int i = 0; i < sortedPositions.length; i++) {
             sortedPositions[i] = i;
         }
 
-        IntArrays.quickSort(sortedPositions, 0, sortedPositions.length, (left, right) -> {
-            boolean nullLeft = elementBlock.isNull(left);
-            boolean nullRight = elementBlock.isNull(right);
-            if (nullLeft && nullRight) {
-                return 0;
-            }
-            if (nullLeft) {
-                return 1;
-            }
-            if (nullRight) {
-                return -1;
-            }
-
-            return elementBlock.compareTo(
-                    left,
-                    0,
-                    elementBlock.getSliceLength(left),
-                    elementBlock,
-                    right,
-                    0,
-                    elementBlock.getSliceLength(right));
-        });
-
+        IntArrays.quickSort(sortedPositions, 0, sortedPositions.length, dictionary::compareIndex);
         return sortedPositions;
     }
 
     @Override
     protected Optional<int[]> writeDictionary()
     {
-        Block dictionaryElements = dictionary.getElementBlock();
-
-        // write dictionary in sorted order
-        int[] sortedDictionaryIndexes = getSortedDictionaryNullsLast(dictionaryElements);
+        int[] sortedDictionaryIndexes = getSortedDictionary(dictionary);
         for (int sortedDictionaryIndex : sortedDictionaryIndexes) {
-            if (!dictionaryElements.isNull(sortedDictionaryIndex)) {
-                int length = dictionaryElements.getSliceLength(sortedDictionaryIndex);
-                dictionaryLengthStream.writeLong(length);
-                Slice value = dictionaryElements.getSlice(sortedDictionaryIndex, 0, length);
-                dictionaryDataStream.writeSlice(value);
-            }
+            int length = dictionary.getSliceLength(sortedDictionaryIndex);
+            dictionaryLengthStream.writeLong(length);
```

Comment:
is this optimal when compared to the previous one?

Reply:
This avoids creating intermediary slice value (temporary object). rawSlice is reference to the same object and this uses offsets to write directly, so it is more efficient.